hips/hip-00XX.md

```diff
@@ -0,0 +1,213 @@
+As SSA allows custom resources to be correctly merged based on metadata within the CRD.
+Or the API server will use the default approach (<https://kubernetes.io/docs/reference/using-api/server-side-apply/#custom-resources>).
+
+#### Conflicts and forcing
+
+It is possible that when Helm upgrades (or rollbacks) a chart, there will be a field conflict with another field manager.
+In this case, Helm will report the error for the conflict field(s) to the user.
+
+Helm will also add a new flag `--force-conflicts=false|true` to the install, upgrade and rollback commands (modeled after the life `kubectl apply` flag)
+And corresponding API objects field.
+When `true` (default `false`), Helm will enabling override the field conflicts, setting field(s) to the charts values.
+And also make Helm the field manager for the conflicting field(s).
+The `--force-conflicts` flag documentation should explicity mention the downsides of conflict forcing ie. the chart update is overriding object values that another process is expecting to own/manage.
+
+For more details, see the _Overwrite value, become sole manager_ of <https://kubernetes.io/docs/reference/using-api/server-side-apply/#conflicts>
+
+(Of note, Helm upgrade already includes the `--force` flag, which tells Helm to replace objects during upgrade.
+However, given this different semantics of `--force`, a new flag is required)
+
+An additional special case is rollbacks on failed upgrade (`--atomic` flag). It is possible for Helm to attempt to rollback to a version which now has a conflict.
+For which the rollback will fail.
+Since rollbacks are "best effort" and can fail for other reasons as well, it seems best to treat this new rollback failure-case similarly (report to the user via logging and error).
+If the user specified `--force-conflicts` to the upgrade, the corresponding rollback should similarly force conflicts (this is the behavior of `--force`).
+If the rollback is to a release version which previously didn't have SSA enabled, the rollback install should also disable SSA.
+
+#### Special
+
+- `--dry-run=client` won't work with SSA enabled
+- If the user attempts to use SSA with a cluster which does not support it (unlikely: SSA [went GA](https://kubernetes.io/blog/2021/08/06/server-side-apply-ga/) in Kubernetes v1.22), Helm should error
+- Client-side API object validation isn't applicable when SSA enabled (`helm install|upgrade --disable-openapi-validation`)
+- Hooks could be deployed with SSA. But as hook objects are not updated (create/delete only), they are not really affected
+- Stored release manifests should not change (they store the object as Helm intends)
+
+### Implementation
+
+Helm will send object's manifests to the Kubernetes API server via SSA patches. Similar to `kubectl apply`.
+Utilizing the field manager name `"helm"`:
```

Comment:
Should this be configurable? I'm thinking of a case where "Sam" and "Alex" might be managing the same namespace and want to be sure that they're not stepping on each other's toes.

Reply:
This HIP is only going as far as Helm supporting a single field manager. For the goal of switching from client-side to server-side apply.

I do think in the future, Helm supporting multiple field managers in the same chart with similar control on conflict resolution, should be done. But that ends up being a can of worm (stuff to design) I think we should defer for the moment.

Reply:
I should mention, see Rejected Ideas for some discussion